### DISCUSSION TITLE
Minimalize dependencies and define scope

### DISCUSSION THREAD

**Discussion by bunny-therapist** [fabiocaccamo/python-benedict#298] [Q&A]
Recently, several dependencies were extracted into "extras", which I think was a great move. I only wish it would go further.

We are currently using benedict exclusively for the keypath functionality and some general dict-utility functionality like "flatten". Because we are using benedict, we get an additional 6 dependencies (benedict has 7, but we already have requests as a dependency). As far as I can tell, most of these dependencies are only used in special utility functions like "get_string" and similar.

To me, the scope of python-benedict is unclear. The dict functionality is very useful (and, I suspect, the main thing people use from benedict?), but there is a plethora of functions like "get_string", "get_slug", "get_phonenumber", "get_datetime", etc, that all seem extremely specific and niche, and add further dependencies. It is not clear to me what the imagined scope of benedict is; theoretically, you could add a "get_<thing>" method for any conceivable object defined in some other library ("get_social_security_number", "get_timestamp", "get_monetary_sum", etc), and it is not clear to me where benedict draws the line. For example, I see issue #249 mentioning support for encrypted data, and I wonder if this will mean that benedict will get additional methods which add more dependencies in the future. Supporting, e.g., parsing of phone numbers in what is to a large extent an extended-dict-handling library already seems to suggest feature creep. As a user, I would have loved it if benedict was just keypath/keyattr/keylist dict with no further dependencies, or if that basic functionality was one library, and then another library with the rest of the functionality was built on top of it.

We are at the point where we are considering developing some alternative to python-benedict simply because of the extra dependencies and us being unsure about the future direction/scope of benedict. Will more dependencies be shifted to extras? Will more dependencies be added to benedict? Is there any limit to the scope of things benedict will support?

<a href="https://dashboard.polar.sh/fabiocaccamo/python-benedict/issues/297">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://dashboard.polar.sh/api/github/fabiocaccamo/python-benedict/issues/297/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://dashboard.polar.sh/api/github/fabiocaccamo/python-benedict/issues/297/pledge.svg">
</picture>
</a>

**Comment by fabiocaccamo** [comment]
@bunny-therapist thank you for the feedback, I really appreciate it and find it very useful.

### About
Here you can read a bit of history/scope of the library: 
https://dev.to/fabiocaccamo/python-benedict-dict-subclass-with-keypath-support-standardized-io-shortcuts-and-many-utilities-50k2

### Requirements
Along the time many features have been added *(some requested by users, other for personal needs)* and consequently also many requirements have been added.

Recently I moved many requirements to `extras` and I like the idea to move also all the remaining requirements to `extras`, the idea is to go **zero-requirements** by default.

### Encryption
In the past I made an encryption utility for a project I worked on and I thought it would have been useful to other users, so I created that issue as "reminder", but I don't know if it will ever see the light. 

Anyway... if some encryption utility will be added in the future, it will be surely an `extra`, so no new requirements will be installed by default, don't worry :)

---

I hope to have clarified all your doubts.

**Reply by bunny-therapist** [reply]
Thank you. If you want to strive toward zero-requirements and will not add any more mandatory dependencies, then I am happy and my team will keep happily using python-benedict. Thank you so much!

**Reply by fabiocaccamo** [reply]
I will surely move these requirements to extras soon, probably in the next version:

- `ftfy`
- `mailchecker`
- `phonenumbers`
- `python-dateutil`


I'm still undecided about these (what do you think?):
- `python-fsutil`
- `python-slugify`
- `requests`

**Reply by bunny-therapist** [reply]
My understanding is that fsutil and requests are only used to load dict from url (it took me some time to figure out the dependency on requests) (if it is used for anything else, I expect it to be some path utility which could have been done using pathlib/os just as well), and that slugify is just used for get_slug.

I would move those to extras as well. "Zero-requirements" should mean zero, and there really is no need to have these extra dependencies just for a function or two.

**Reply by fabiocaccamo** [reply]
- `python-fsutil` is used by all `from_{format}` and `to_{format}` methods when doing file-system operations, some of these methods (eg. `from_json` and `to_json`) work even without installing extras.
- `python-slugify` is used by the `standardize` and the `get_slug` methods.

**Reply by bunny-therapist** [reply]
I don't personally have any usecase for slugify.

As for python-fsutil for file-system operations, it seems like that wouldn't really need any extra library at all? I haven't looked at that code, but it sounds like something I would just use pathlib and built-ins for.

(Still, I will say that fsutil makes a lot more sense as a dependency than e.g. phonenumbers.)

**Comment by fabiocaccamo** [comment]
@bunny-therapist there is also another discussion I started some time ago but nobody partecipated: https://github.com/fabiocaccamo/python-benedict/discussions/116

**Reply by bunny-therapist** [reply]
Yes, I know, but it was locked to me. I checked all issues and discussions before making an issue, but this seemed to be the only way I could communicate.

**Reply by fabiocaccamo** [reply]
Ah yes, probably I closed it after I added extra requirements, I unlock it. (no problem for having opened an issue)
